pkg/ccl/backupccl/split_and_scatter_processor.go

```diff
@@ -0,0 +1,280 @@
+	rangeInfos := res.Header().RangeInfos
+	if len(rangeInfos) == 1 {
+		destination = rangeInfos[0].Lease.Replica.NodeID
+	} else {
+		log.Warningf(ctx, "scatter request unexpectedly scattered %d ranges", len(rangeInfos))
+	}
+	return destination, nil
+}
+
+var splitAndScatterOutputTypes = []*types.T{
+	types.Bytes, // Span key for the range router
+	types.Bytes, // RestoreDataEntry bytes
+}
+
+// splitAndScatterProcessor is given a set of spans (specified as
+// RestoreSpanEntry's) to distribute across the cluster. Depending on which node
+// the span ends up on, it forwards RestoreSpanEntry as bytes along with the key
+// of the span on a row. It expects an output RangeRouter and before it emits
+// each row, it updates the entry in the RangeRouter's map with the destination
+// of the scatter.
+type splitAndScatterProcessor struct {
+	flowCtx   *execinfra.FlowCtx
+	spec      execinfrapb.SplitAndScatterSpec
+	output    execinfra.RowReceiver
+	scatterer splitAndScatterer
+}
+
+var _ execinfra.Processor = &splitAndScatterProcessor{}
+
+// OutputTypes implements the execinfra.Processor interface.
+func (ssp *splitAndScatterProcessor) OutputTypes() []*types.T {
+	return splitAndScatterOutputTypes
+}
+
+func newSplitAndScatterProcessor(
+	flowCtx *execinfra.FlowCtx,
+	_ int32,
+	spec execinfrapb.SplitAndScatterSpec,
+	output execinfra.RowReceiver,
+) (execinfra.Processor, error) {
+	ssp := &splitAndScatterProcessor{
+		flowCtx:   flowCtx,
+		spec:      spec,
+		output:    output,
+		scatterer: dbSplitAndScatterer{},
+	}
+	return ssp, nil
+}
+
+type entryNode struct {
+	entry execinfrapb.RestoreSpanEntry
+	node  roachpb.NodeID
+}
+
+// Run implements the execinfra.Processor interface.
+func (ssp *splitAndScatterProcessor) Run(ctx context.Context) {
+	ctx, span := tracing.ChildSpan(ctx, "splitAndScatterProcessor")
+	defer tracing.FinishSpan(span)
+	defer ssp.output.ProducerDone()
+
+	numEntries := 0
+	for _, chunk := range ssp.spec.Chunks {
+		numEntries += len(chunk.Entries)
+	}
+	// Large enough so that it never blocks.
+	doneScatterCh := make(chan entryNode, numEntries)
+
+	var err error
+	splitAndScatterCtx, cancelSplitAndScatter := context.WithCancel(ctx)
+	defer cancelSplitAndScatter()
+	// Note that the loop over doneScatterCh should prevent this goroutine from
+	// leaking when there are no errors. However, if that loop needs to exit
+	// early, runSplitAndScatter's context will be canceled.
+	go func() {
+		defer close(doneScatterCh)
+		err = runSplitAndScatter(splitAndScatterCtx, ssp.flowCtx, &ssp.spec, ssp.scatterer, doneScatterCh)
+		if err != nil {
+			log.Errorf(ctx, "error while running split and scatter: %+v", err)
+		}
+	}()
+
+	for scatteredEntry := range doneScatterCh {
+		entry := scatteredEntry.entry
+		entryBytes, err := protoutil.Marshal(&entry)
+		if err != nil {
+			ssp.output.Push(nil, &execinfrapb.ProducerMetadata{Err: err})
+			break
+		}
+
+		routingDatum, _ := routingDatumsForNode(scatteredEntry.node)
```

Comment:
can easily be a follow-up, but we could trivially memoize this in a map by nodeid to avoid the alloc of a new routing datum every time, right?

Reply:
Yep, added a cache.